File: src/lib/WritableToCSVStream.java
package lib;

import java.io.IOException;

public interface WritableToCSVStream {
    /**
     * If object is null writes null to stream, otherwise calls the function
     *
     * @param function, which will be called in case object is not null
     */
    static <T> void writeNullableToStream(
            CSVStreamWriter stream,
            T object,
            FunctionWithVoidReturnOneArgumentAndError<T, IOException> function
    ) throws IOException {
        if (object == null) {
            stream.append("null");
        } else {
            function.invoke(object);
        }
    }

    void writeToStream(CSVStreamWriter stream) throws IOException;
}

File: src/lib/WritableToCSVStream.kt
package lib

import lib.CSV.CSVStreamWriter
import lib.functions.FunctionWithVoidReturnAndOneArgument

/**
 * If object is null writes null to stream, otherwise calls the function
 *
 * @param function, which will be called in case object is not null
 */
fun <T> writeNullableToStream(
    stream: CSVStreamWriter,
    value: T?,
    function: FunctionWithVoidReturnAndOneArgument<T>
) {
    if (value == null) {
        stream.append("null")
    } else {
        function.invoke(value)
    }
}

fun interface WritableToCSVStream {
    fun writeToStream(stream: CSVStreamWriter)
}
